### PR TITLE
[eda] Feature Interaction - Feature Distance

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*

# Feature Distance Analysis

[Open In Colab](https://colab.research.google.com/drive/1Oecs3fBI5QqHCMwjl34ZylAy1oVE5REm?usp=sharing)

The component performs feature correlation distance analysis using Spearman rank correlation and hierarchical clustering for the data passed in `train_data` excluding `label`. The near duplicates grouping is automatically suggested given `near_duplicates_threshold`.The results can be visualized using `FeatureDistanceAnalysisVisualization`.

Note: it is recommended to apply `ApplyFeatureGenerator` before the analysis to ensure correlations are calculated for categorical variables.

### Parameters

```
Parameters
----------
near_duplicates_threshold: float, default = 0.01
    defines feature distance to be considered as near duplicates
```

Note: this analysis is an augmentation for `FeatureInteraction` and should be used in pair to be visualized via `FeatureInteractionVisualization`.

### Example

```python
import autogluon.eda.analysis as eda
import autogluon.eda.visualization as viz
import autogluon.eda.auto as auto
import pandas as pd
import numpy as np
 
df_train = pd.DataFrame(...)
 
auto.analyze(
    train_data=df_train, label=target_col,
    anlz_facets=[
        eda.transform.ApplyFeatureGenerator(category_to_numbers=True, children=[
            eda.interaction.FeatureDistanceAnalysis(near_duplicates_threshold=0.7),
        ])
    ],
    viz_facets=[
        viz.layouts.MarkdownSectionComponent('### Feature distance analysis'),
        viz.interaction.FeatureDistanceAnalysisVisualization(fig_args=dict(figsize=(12,6))),
    ]
)
```

image

Credit: See [Removing-redundant-features](https://github.com/fastai/fastbook/blob/master/09_tabular.ipynb) section in Jeremy Howard's "Deep Learning for Coders with Fastai and PyTorch" book.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: eda/src/autogluon/eda/analysis/transform.py [hunk 1]
```diff
@@ -60,15 +60,18 @@ def __init__(
...
                 enable_numeric_features=True,
                 enable_categorical_features=True,
                 enable_datetime_features=False,
                 enable_text_special_features=False,
                 enable_text_ngram_features=False,
                 enable_raw_text_features=False,
                 enable_vision_features=False,
+                verbosity=0,
             )
+            args = {**args, **kwargs}
```
Comment (gidler) [inline eda/src/autogluon/eda/analysis/transform.py:73]: Is this more readable:
```
            feature_generator = AutoMLPipelineFeatureGenerator(
                enable_numeric_features=True,
                enable_categorical_features=True,
                enable_datetime_features=False,
                enable_text_special_features=False,
                enable_text_ngram_features=False,
                enable_raw_text_features=False,
                enable_vision_features=False,
                verbosity=0,
                **kwargs,
            )
```

File: eda/tests/unittests/analysis/test_interaction.py [hunk 2]
```diff
@@ -1,12 +1,34 @@
 from unittest.mock import MagicMock
 
 import pandas as pd
+import numpy as np
 import pytest
+import os
```
Comment (gidler) [inline eda/tests/unittests/analysis/test_interaction.py:6]: Not a big deal, but import order is weird here

**Review by gidler — APPROVED** [review]
Approved with a couple small comments

File: eda/tests/unittests/analysis/test_interaction.py (hunk 2 above)
Comment (yinweisu) [reply eda/tests/unittests/analysis/test_interaction.py:6]: BTW, just realized that lint check are not enabled for EDA. Is this intended? I somehow remember you've enabled it @gradientsky

**Review by yinweisu — APPROVED** [review]

File: eda/src/autogluon/eda/analysis/transform.py (hunk 1 above)
Comment (gradientsky) [reply eda/src/autogluon/eda/analysis/transform.py:73]: updated

File: eda/tests/unittests/analysis/test_interaction.py (hunk 2 above)
Comment (gradientsky) [reply eda/tests/unittests/analysis/test_interaction.py:6]: there is a lint check: https://github.com/autogluon/autogluon/blob/master/.github/workflow_scripts/test_eda.sh#L14

**Pull request merged**